File: internal/metrics/metrics.go
package metrics

import (
	"bufio"
	"encoding/csv"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"strconv"
	"strings"
	"time"

	"github.com/kgsaran/trackfw/internal/pathguard"
)

// Transition representa uma entrada do .trackfw-log.
type Transition struct {
	Timestamp time.Time
	Basename  string
	From, To  string
}

// WIPEntry representa um roadmap atualmente em wip com sua idade.
type WIPEntry struct {
	Basename string
	Age      time.Duration
}

// Metrics contém as métricas de delivery calculadas.
type Metrics struct {
	CycleTimeMean time.Duration
	Throughput    float64 // roadmaps concluídos por semana
	WIPEntries    []WIPEntry
	Period        string
}

// lineRe faz match de linhas do formato:
// 2026-06-12 14:30  ROADMAP-2026-06-12-auth.md                  backlog → wip
var lineRe = regexp.MustCompile(`^(\d{4}-\d{2}-\d{2} \d{2}:\d{2})\s{2,}(\S+)\s{2,}(\S+)\s+→\s+(\S+)`)

// metricsGetwdFn is the function used to obtain the current working directory inside
// ExportCSV. It defaults to os.Getwd and may be overridden in tests to inject a
// controlled failure without relying on OS-specific filesystem tricks.
var metricsGetwdFn func() (string, error) = os.Getwd

// ParseLog lê o arquivo .trackfw-log e retorna todas as transições.
// Retorna nil, nil se o arquivo não existe.
func ParseLog(path string) ([]Transition, error) {
	f, err := os.Open(path)
	if os.IsNotExist(err) {
		return nil, nil
	}
	if err != nil {
		return nil, fmt.Errorf("metrics: abrir log: %w", err)
	}
	defer f.Close()

	var transitions []Transition
	scanner := bufio.NewScanner(f)
	for scanner.Scan() {
		line := scanner.Text()
		if strings.TrimSpace(line) == "" {
			continue
		}
		m := lineRe.FindStringSubmatch(line)
		if m == nil {
			continue
		}
		ts, err := time.ParseInLocation("2006-01-02 15:04", m[1], time.Local)
		if err != nil {
			continue
		}
		transitions = append(transitions, Transition{
			Timestamp: ts,
			Basename:  strings.TrimSpace(m[2]),
			From:      strings.TrimSpace(m[3]),
			To:        strings.TrimSpace(m[4]),
		})
	}
	if err := scanner.Err(); err != nil {
		return nil, fmt.Errorf("metrics: ler log: %w", err)
	}
	return transitions, nil
}

// Filter retorna apenas as transições com Timestamp >= since.
func Filter(transitions []Transition, since time.Time) []Transition {
	var out []Transition
	for _, t := range transitions {
		if !t.Timestamp.Before(since) {
			out = append(out, t)
		}
	}
	return out
}

// Calculate computa cycle time médio, throughput e WIP age a partir das transições.
func Calculate(transitions []Transition) Metrics {
	// Agrupar transições por basename para calcular ciclos.
	type stateEntry struct {
		ts    time.Time
		state string
	}
	// Mapa: basename → slice de entradas em ordem cronológica
	byName := make(map[string][]stateEntry)
	for _, t := range transitions {
		byName[t.Basename] = append(byName[t.Basename], stateEntry{ts: t.Timestamp, state: t.To})
	}

	// Calcular cycle time: tempo desde entrada em backlog (ou wip) até done.
	var cycleTimes []time.Duration
	for _, entries := range byName {
		var startTs *time.Time
		var doneTs *time.Time
		for _, e := range entries {
			if (e.state == "backlog" || e.state == "wip") && startTs == nil {
				ts := e.ts
				startTs = &ts
			}
			if e.state == "done" {
				ts := e.ts
				doneTs = &ts
			}
		}
		if startTs != nil && doneTs != nil {
			cycleTimes = append(cycleTimes, doneTs.Sub(*startTs))
		}
	}

	var cycleTimeMean time.Duration
	if len(cycleTimes) > 0 {
		var total time.Duration
		for _, ct := range cycleTimes {
			total += ct
		}
		cycleTimeMean = total / time.Duration(len(cycleTimes))
	}

	// Calcular throughput: roadmaps concluídos por semana.
	var doneCount int
	var minTs, maxTs time.Time
	for _, t := range transitions {
		if t.To == "done" {
			doneCount++
		}
		if minTs.IsZero() || t.Timestamp.Before(minTs) {
			minTs = t.Timestamp
		}
		if maxTs.IsZero() || t.Timestamp.After(maxTs) {
			maxTs = t.Timestamp
		}
	}
	var throughput float64
	if doneCount > 0 {
		weeks := maxTs.Sub(minTs).Hours() / (7 * 24)
		if weeks < 1 {
			weeks = 1
		}
		throughput = float64(doneCount) / weeks
	}

	// Calcular WIP age: basenames que entraram em wip mas não foram concluídos ou abandonados.
	var wipEntries []WIPEntry
	for name, entries := range byName {
		var wipTs *time.Time
		concluded := false
		for _, e := range entries {
			if e.state == "wip" {
				ts := e.ts
				wipTs = &ts
			}
			if e.state == "done" || e.state == "abandoned" {
				concluded = true
			}
		}
		if wipTs != nil && !concluded {
			wipEntries = append(wipEntries, WIPEntry{
				Basename: name,
				Age:      time.Since(*wipTs),
			})
		}
	}

	return Metrics{
		CycleTimeMean: cycleTimeMean,
		Throughput:    throughput,
		WIPEntries:    wipEntries,
	}
}

// ExportCSV grava as transições e métricas em um arquivo CSV.
//
// Containment guard: when path is relative (or is absolute but contained inside
// the project root), RejectSymlinks is applied to reject any symlink ancestor
// between the project root and the destination.
//
// Named exception: an absolute path that resolves outside the project root
// (e.g. /tmp/metrics.csv) is user-directed — the user explicitly chose an
// external destination. That case is not "derived from root" per
// ADR-2026-09-18 and is allowed through without a guard.
func ExportCSV(m Metrics, transitions []Transition, path string) error {
	// Resolve project root from CWD — macOS /tmp→/private/tmp invariant.
	// Fail closed: if Getwd() fails we cannot determine whether the path is inside
	// the project root, so we refuse the write rather than proceeding unguarded.
	cwd, cwdErr := metricsGetwdFn()
	if cwdErr != nil {
		return pathguard.RefuseUnverifiableRoot(path, cwdErr)
	}
	exportRoot := cwd
	if resolved, resolveErr := filepath.EvalSymlinks(cwd); resolveErr == nil {
		exportRoot = resolved
	}
	absPath := path
	if !filepath.IsAbs(path) {
		absPath = filepath.Join(exportRoot, path)
	}
	// Guard only paths inside the project root. External absolute paths are a
	// named exception: user-directed, not derived from root (ADR-2026-09-18).
	// When Beneath returns false (external absolute path), the guard is intentionally
	// skipped — the caller explicitly chose an external destination.
	if pathguard.Beneath(exportRoot, absPath) {
		if guardErr := pathguard.RejectAndReport(exportRoot, absPath); guardErr != nil {
			return guardErr
		}
	}
	// Guarded when path is beneath project root (RejectSymlinks above); external absolute paths are
	// a user-directed named exception per ADR-2026-09-18 (Beneath returns false → guard skipped).
	// Fail closed on Getwd() error (early return above prevents reaching this line in that case).
	// write-containment-allowed: guarded by pathguard.RejectSymlinks above; Beneath-false paths are user-directed named exception (ADR-2026-09-18)
	f, err := os.Create(path)
	if err != nil {
		return fmt.Errorf("metrics: criar CSV: %w", err)
	}
	defer f.Close()

	w := csv.NewWriter(f)

	// Seção de transições.
	if err := w.Write([]string{"basename", "from", "to", "timestamp"}); err != nil {
		return fmt.Errorf("metrics: escrever header CSV: %w", err)
	}
	for _, t := range transitions {
		if err := w.Write([]string{
			t.Basename,
			t.From,
			t.To,
			t.Timestamp.Format("2006-01-02 15:04"),
		}); err != nil {
			return fmt.Errorf("metrics: escrever linha CSV: %w", err)
		}
	}

	// Linha em branco.
	if err := w.Write([]string{""}); err != nil {
		return fmt.Errorf("metrics: escrever separador CSV: %w", err)
	}

	// Seção de métricas.
	if err := w.Write([]string{"metric", "value"}); err != nil {
		return fmt.Errorf("metrics: escrever header métricas CSV: %w", err)
	}
	cycleHours := m.CycleTimeMean.Hours()
	if err := w.Write([]string{"cycle_time_mean_hours", strconv.FormatFloat(cycleHours, 'f', 2, 64)}); err != nil {
		return fmt.Errorf("metrics: escrever cycle_time: %w", err)
	}
	if err := w.Write([]string{"throughput_per_week", strconv.FormatFloat(m.Throughput, 'f', 2, 64)}); err != nil {
		return fmt.Errorf("metrics: escrever throughput: %w", err)
	}
	if err := w.Write([]string{"wip_count", strconv.Itoa(len(m.WIPEntries))}); err != nil {
		return fmt.Errorf("metrics: escrever wip_count: %w", err)
	}

	w.Flush()
	return w.Error()
}

File: pypi/trackfw/commands/metrics.py
"""
commands/metrics.py — Subcomando `trackfw metrics`.
Lê .trackfw-log e calcula throughput, cycle time e WIP age.
Espelho Python de npm/src/commands/metrics.js.
"""

import os
import re
import sys
from datetime import datetime, timedelta

# Regex que faz match do formato do .trackfw-log
# 2026-06-12 14:30  ROADMAP-2026-06-12-auth.md                  backlog → wip
LINE_RE = re.compile(
    r"^(\d{4}-\d{2}-\d{2} \d{2}:\d{2})\s{2,}(\S+)\s{2,}(\S+)\s+→\s+(\S+)"
)

LOG_PATH = os.path.join("docs", "roadmaps", ".trackfw-log")


def _parse_log(file_path):
    """
    Lê o arquivo .trackfw-log e retorna lista de transições.
    Retorna [] se o arquivo não existe.
    """
    if not os.path.exists(file_path):
        return []
    transitions = []
    try:
        with open(file_path, "r", encoding="utf-8") as f:
            content = f.read()
    except OSError:
        return []
    for line in content.split("\n"):
        line = line.strip()
        if not line:
            continue
        m = LINE_RE.match(line)
        if not m:
            continue
        ts_str, basename, from_state, to_state = m.group(1), m.group(2), m.group(3), m.group(4)
        try:
            timestamp = datetime.strptime(ts_str, "%Y-%m-%d %H:%M")
        except ValueError:
            continue
        transitions.append({
            "timestamp": timestamp,
            "basename": basename.strip(),
            "from": from_state.strip(),
            "to": to_state.strip(),
        })
    return transitions


def _filter(transitions, since_dt):
    """Retorna transições com timestamp >= since_dt."""
    return [t for t in transitions if t["timestamp"] >= since_dt]


def _calculate(transitions):
    """
    Computa cycle time médio, throughput e WIP age.
    Retorna dict {cycle_time_mean_s, throughput, wip_entries}.
    """
    by_name = {}
    for tr in transitions:
        name = tr["basename"]
        by_name.setdefault(name, []).append(tr)

    # Cycle time: da entrada em backlog ou wip até done
    cycle_times = []
    for entries in by_name.values():
        start_ts = None
        done_ts = None
        for e in entries:
            if e["to"] in ("backlog", "wip") and start_ts is None:
                start_ts = e["timestamp"]
            if e["to"] == "done":
                done_ts = e["timestamp"]
        if start_ts is not None and done_ts is not None:
            delta = (done_ts - start_ts).total_seconds()
            cycle_times.append(delta)

    cycle_time_mean_s = 0.0
    if cycle_times:
        cycle_time_mean_s = sum(cycle_times) / len(cycle_times)

    # Throughput: roadmaps done por semana
    done_count = 0
    timestamps = [tr["timestamp"] for tr in transitions]
    for tr in transitions:
        if tr["to"] == "done":
            done_count += 1

    throughput = 0.0
    if done_count > 0 and timestamps:
        min_ts = min(timestamps)
        max_ts = max(timestamps)
        delta_s = (max_ts - min_ts).total_seconds()
        weeks = delta_s / (7 * 24 * 3600)
        if weeks < 1:
            weeks = 1
        throughput = done_count / weeks

    # WIP age: basenames em wip sem done ou abandoned posterior
    now = datetime.now()
    wip_entries = []
    for basename, entries in by_name.items():
        wip_ts = None
        concluded = False
        for e in entries:
            if e["to"] == "wip":
                wip_ts = e["timestamp"]
            if e["to"] in ("done", "abandoned"):
                concluded = True
        if wip_ts is not None and not concluded:
            age_s = (now - wip_ts).total_seconds()
            wip_entries.append({"basename": basename, "age_s": age_s})

    return {
        "cycle_time_mean_s": cycle_time_mean_s,
        "throughput": throughput,
        "wip_entries": wip_entries,
    }


def _format_duration(seconds):
    """Formata segundos em string legível (days/hours)."""
    total_hours = int(seconds / 3600)
    days = total_hours // 24
    hours = total_hours % 24
    if days > 0:
        return f"{days} days {hours} hours"
    return f"{hours} hours"


def _export_csv(metrics, transitions, file_path):
    """Grava transições e métricas em um arquivo CSV."""
    rows = ["basename,from,to,timestamp"]
    for tr in transitions:
        ts = tr["timestamp"].strftime("%Y-%m-%d %H:%M")
        rows.append(f"{tr['basename']},{tr['from']},{tr['to']},{ts}")
    rows.append("")
    rows.append("metric,value")
    cycle_hours = metrics["cycle_time_mean_s"] / 3600
    rows.append(f"cycle_time_mean_hours,{cycle_hours:.2f}")
    rows.append(f"throughput_per_week,{metrics['throughput']:.2f}")
    rows.append(f"wip_count,{len(metrics['wip_entries'])}")
    with open(file_path, "w", encoding="utf-8") as f:
        f.write("\n".join(rows) + "\n")


def _print_metrics(metrics):
    """Imprime as métricas em formato tabela ASCII."""
    print("── trackfw metrics ──────────────────────")

    if metrics["cycle_time_mean_s"] > 0:
        print(f"  Cycle Time Mean   : {_format_duration(metrics['cycle_time_mean_s'])}")
    else:
        print("  Cycle Time Mean   : n/a (no completed cycles)")

    if metrics["throughput"] > 0:
        print(f"  Throughput        : {metrics['throughput']:.2f} roadmaps/week")
    else:
        print("  Throughput        : n/a (no completed roadmaps)")

    wip = metrics["wip_entries"]
    if not wip:
        print("  WIP Age           : no items in progress")
    else:
        print(f"  WIP Age ({len(wip)} items) :")
        for w in wip:
            print(f"    - {w['basename']}: {_format_duration(w['age_s'])}")

    print("─────────────────────────────────────────")


def _parse_days(s):
    """Converte '7d', '30d', '90d' em número de dias (int)."""
    s = s.strip()
    if not s or s[-1] != "d":
        raise ValueError(f"formato inválido: '{s}' (use: 7d, 30d, 90d)")
    try:
        n = int(s[:-1])
    except ValueError:
        raise ValueError(f"número inválido em '{s}'")
    if n <= 0:
        raise ValueError(f"número deve ser positivo: '{s}'")
    return n


def register(subparsers):
    """Adiciona subcomando `metrics` ao parser principal."""
    parser = subparsers.add_parser(
        "metrics",
        help="Show delivery metrics from .trackfw-log",
    )
    parser.add_argument(
        "--days",
        metavar="N",
        default=None,
        help="Filter to last N days (e.g. --days 30)",
    )
    parser.add_argument(
        "--since",
        metavar="PERIOD",
        default=None,
        help="Filter period in format Nd (e.g. --since 30d)",
    )
    parser.add_argument(
        "--export",
        metavar="FILE",
        default=None,
        help="Export metrics to CSV file",
    )
    parser.set_defaults(func=_cmd_metrics)


def _cmd_metrics(args):
    transitions = _parse_log(LOG_PATH)

    if not transitions:
        print("No log found")
        return

    # Suporte a --days N e --since Nd
    since_dt = None
    if getattr(args, "days", None):
        try:
            n = int(args.days)
            since_dt = datetime.now() - timedelta(days=n)
        except ValueError:
            print(f"invalid --days value: {args.days}", file=sys.stderr)
            sys.exit(1)
    elif getattr(args, "since", None):
        try:
            n = _parse_days(args.since)
            since_dt = datetime.now() - timedelta(days=n)
        except ValueError as e:
            print(f"invalid --since format: {e}", file=sys.stderr)
            sys.exit(1)

    if since_dt is not None:
        transitions = _filter(transitions, since_dt)
        if not transitions:
            print("No log found")
            return

    metrics = _calculate(transitions)
    _print_metrics(metrics)

    if getattr(args, "export", None):
        _export_csv(metrics, transitions, args.export)
        print(f"exported to {args.export}")
